mgr/dashboard_v2: Add table component

Devp00l opened the pull request:
The current implementation isn't finished yet, but it's a start.

The ngx-datatable is used as the core of the table component. You can use the
table to show data in a table. You can search, paginate, sort, refresh the
table contents. Multi selection is possible, the details of the selected items
will be given to the specified detail component. The table is styled in a
similar way as the old oA table.

What will be fixed soon?
* Enable the usage of buttons in the table header
* Enable the possibility to show the details inline not beneath the table
* The footer will be replaced to look as in the old oA table
* The columns to show can be checked and predefined
* The selection made by the user will be saved in the local storage

tspmelo reviewed (changes requested):
We should pass a fetch method to the oa-table so it can execute it when needed.
This method should include a call to the "REST service" and process the results.

tspmelo reviewed (changes requested):
We should use a column input in oa-table, that will contains all the column definitions and which template each one should use.
This is not the same as the column attribute that ngx-datatable uses.

tspmelo reviewed (changes requested):
Remove the "data" input.

LenzGr reviewed (dismissed):
lgtm, thanks

Devp00l commented:
Updated the commit

Now the data table is usable by any component.
The table can now trigger the get function of the callee to update the
data.
It can now load the details component if added as an
"entryComponent" to the "components" module. The template will get an
array of the selected items.
Column definitions can be passed to the table.

TODO:
* Add interface for column definitions and details components
* Fix some style issues
* Make it possible to load the template under each selected entry
* Fix all tests


An example can be viewed here, how to use it, as there will be no mayor changes in the usage anymore:
[sample-osd-page](https://github.com/Devp00l/ceph/tree/data-table-osd-sample)

jecluis reviewed (changes requested):
This patch commit has a `Signed-off-by` for @tspmelo in the beginning of the commit message's body. Please move it to the end of the commit message, just below (or above) the @Devp00l `Signed-off-by`.

Commit:
mgr/dashboard_v2: Add table component

The current implementation isn't finished yet, but it's a start.

The ngx-datatable is used as the core of the table component. You can
use the table to show data in a table. You can search, paginate, sort,
refresh the table contents. Multi selection is possible, the details
of the selected items will be given to the specified detail component.

What will be fixed soon?
* Enable the usage of buttons in the table header
* Enable details inline not beneath the table
* Pagination to use a input field to switch pages
* The columns to show can be checked and predefined
* The selection made by the user will be saved in the local storage

Signed-off-by: Stephan Müller <smueller@suse.com>
Signed-off-by: Tiago Melo <tmelo@suse.com>

Devp00l commented:
Just to remind you that it only has the basic functionality, but we will make it awesome ;)

jecluis commented:
@Devp00l am I recalling this incorrectly, or has your commit gone from actually having information in its body to having no useful information at all? Also, please amend your commit title (at least before merging) so it reflects the component being changed.

jecluis commented:
Additionally, if the commit is the result of squashing multiple authors' commits, you should consider adding the other authors' `Signed-off-by` as well.

Exotelis reviewed (changes requested):
Consider nesting the selectors in your scss files and using mixins if possible

Devp00l commented:
![peek 2018-02-02 15-05](https://user-images.githubusercontent.com/16167865/35737773-f9b4ad80-082c-11e8-8d27-6d338e5fa6d3.gif)

Devp00l commented:
Fixed everything.

Exotelis reviewed (dismissed):
I don't like the "new" pagination. Because if you have e.g. 50 pages you have to click too often to get to page 25. Would it be possible to change this with the lib.?

Beside this it lgtm

jecluis commented:
@Devp00l do you mind adjusting your commit message according to the 50/72 rule? i.e., 50 columns max for the title (which you already do), 72 columns max for the body (with you don't). Typically, vim will do that automatically if using it as the default commit editor (at least it does for all of my deployments). Github will also show weird commit messages if you go over those limits, as it will hard-wrap commit messages. Besides, I think pretty much everyone tends to adhere to this convention, probably because it's somewhat encouraged by the kernel community.

votdev commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.ts:
```diff
@@ -0,0 +1,103 @@
+  @Input() detailsComponent?: string; // name of the component fe 'TableDetailsComponent'
+  @Input() header? = true;
+
+  @Output() fetchData = new EventEmitter(); // Should be the function that will update the input data
```
Wouldn't it make sense to declare it as ``@Output() fetchData = new EventEmitter<string>();``. This way it is clear that a string is submitted as function argument by the ``emit`` function.

Devp00l replied:
Changed the usage of "fetchData" to not call  it with a string

votdev commented on src/pybind/mgr/dashboard_v2/frontend/src/defaults.scss:
```diff
@@ -1 +1,10 @@
 $oa-color-blue: #288cea;
+$oa-color-light-blue: #afd9ee;
+$bg-light-blue: #d9edf7;
```
I suggest  to use ``$bg-color-light-blue`` or ``$bg-light-blue-color`` to make it clear that this is a color.

Devp00l commented:
@Exotelis This is only the default pagination of the "ngx-datatable" it going to be replaced with a similar one that is used in oA 3.x in the future

Fixed any other issue so far.

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/components.module.ts:
```diff
@@ -0,0 +1,16 @@
+import { NgxDatatableModule } from '@swimlane/ngx-datatable';
+import { HttpClientModule, HttpClient } from '@angular/common/http';
+import { TableDetailsDirective } from './table/table-details.directive';
+import { PaginationComponent } from './table/pagination/pagination.component';
```
I'm having a compilation error because this component does not exist

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/components.module.ts:
```diff
@@ -0,0 +1,16 @@
+import {FormsModule} from '@angular/forms';
+
+@NgModule({
+  entryComponents: [OsdDetailsComponent],
```
Another compilation error here, `OsdDetailsComponent` does not exist.

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.html:
```diff
@@ -0,0 +1,77 @@
+  </ngx-datatable>
+  <!--
+                 [offset]="offset"
+  <cd-pagination (currentPage)="offset" [entries]="rows" [limit]="filterConfig.entries"></cd-pagination> -->
```
Comments like this one, as well as unused code should be removed.

ricardoasmarques reviewed (changes requested):
Please remove unused code and commented code.

After that, LGTM.

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.ts:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  ngOnChanges(changes) {
+    if (changes.data.firstChange) {
```
Do you really need this `if`?

On my local environment this is preventing the display of the data that is received from a server request.

Devp00l replied:
Yeah this could be the case if data is "undefined" beforehand, didn't had that case in mind, thanks for spotting this.

ricardoasmarques commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.ts:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  useData() {
+    if (this.data.length === 0) {
```
Should we keep this `if`?

It will prevent the update when we delete all row's, or am I missing something?

Devp00l replied:
Removed it. Thanks for spotting it.

ricardoasmarques reviewed (dismissed):
lgtm

jecluis reviewed (changes requested):
I'd suggest removing this statement from the commit message:

> The table is styled in a similar way as the old oA table.

Reasoning being that those reviewing this commit will likely not be familiar with what "the old oA table" looks like.

Same reasoning can be applied to

> The footer will be replaced to look as in the old oA table

Would be preferable to tell us what it looks like. I'd much prefer if the "TODO"-kind of list at the end to focus on high-level details instead, though. The other items on the list already do that; this one, however, feels way too specific to be in the list.

Devp00l commented:
@jecluis The commit message should be more self explaining now

tspmelo commented:
Please fix the lint errors.

tspmelo commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.ts:
```diff
@@ -0,0 +1,100 @@
+
+  @Output() fetchData = new EventEmitter(); // Should be the function that will update the input data
+
+  selectAble: String = undefined;
```
s/selectAble/selectable

tspmelo commented on src/pybind/mgr/dashboard_v2/frontend/src/app/shared/components/table/table.component.ts:
```diff
@@ -0,0 +1,100 @@
+
+  useData() {
+    this.rows = [...this.data];
+    this.temp = [...this.rows];
```
Since `this.temp` will always have the same value`this.data` we can just remove it and simply use  `this.data` in `updateFilter`
